src/paginate.jl

```diff
@@ -0,0 +1,185 @@
+
+    layers = if valid(layout_data, layout)
+        paginate_layer(layer, layout_data => layout)
+    elseif valid(row_data, row) && valid(col_data, col)
+        paginate_layer(layer, row_data => row, col_data => col)
+    elseif valid(row_data, row)
+        paginate_layer(layer, row_data => row)
+    elseif valid(col, col)
+        paginate_layer(layer, col_data => col)
+    else
+        [layer]
+    end
+    return layers
+end
+
+function vecs_view(d::Dictionary, indices::AbstractVector)
+    map(d) do value
+        view(value, indices)
+    end
+end
+function vecs_view(vec_of_vecs::AbstractVector, indices::AbstractVector)
+    [view(vec, indices) for vec in vec_of_vecs]
+end
+
+function paginate_layer(layer::ProcessedLayer, data_limiter_pairs::Pair...)
+    subsets = map(Base.splat(getsubsets), data_limiter_pairs)
+    layers = map(Iterators.product(subsets...)) do (subs...,)
+        view_vector = map(all, zip(subs...))
+        primary = vecs_view(layer.primary, view_vector)
+        positional = vecs_view(layer.positional, view_vector)
+        named = vecs_view(layer.named, view_vector)
+        ProcessedLayer(layer; primary, positional, named)
+    end
+    return vec(layers)
+end
+
+"""
+    paginate(l; layout=nothing, row=nothing, col=nothing)
+
+Paginate `l`, the `Layer` or `Layers` object created by an `AlgebraOfGraphics` spec, to create a
+`PaginatedLayers` object.
+This contains a vector of layers where each layer operates on a subset of the input data.
+
+The `PaginatedLayers` object can be passed to `draw` which will return a `Vector{FigureGrid}`
+rather than a single figure.
+
+The keywords that limit the number of subplots on each page are the same that are used
+to specify facets in `mapping`:
+  - `layout`: Maximum number of subplots in a wrapped linear layout.
+  - `row`: Maximum number of rows in a 2D layout.
+  - `col`: Maximum number of columns in a 2D layout.
+
+## Example
+
+```
+d = data((
+    x = rand(1000),
+    y = rand(1000),
+    group1 = rand(string.('a':'i'), 1000),
+    group2 = rand(string.('j':'r'), 1000),
+))
+
+layer_1 = d * mapping(:x, :y, layout = :group1) * visual(Scatter)
+paginated_1 = paginate(layer_1, layout = 9)
+figuregrids = draw(paginated_1)
+
+layer_2 = d * mapping(:x, :y, row = :group1, col = :group2) * visual(Scatter)
+paginated_2 = paginate(layer_2, row = 4, col = 3)
+figuregrid = draw(paginated_2, 1) # draw only the first grid
+```
+"""
+function paginate(
+        l::Union{Layer, Layers};
+        layout = nothing,
+        row = nothing,
+        col = nothing,
+    )
+    layers = l isa Layer ? Layers([l]) : l
+    processed_layers = map(ProcessedLayer, layers.layers)
+    inverted = [paginate_layer(processed_layer; layout, row, col) for processed_layer in processed_layers]
+    layers = map(ProcessedLayers, invert(inverted))
+    return PaginatedLayers(layers, layout, row, col)
+end
+
+# copied from SplitApplyCombine.jl to avoid full dependency 
```

Comment:
I understand that the fully optimized implementation is a bit complex, but couldn't we just do

```julia
function invert(a::AbstractArray{<:AbstractArray})
    f = first(a)
    all(axes(v) == axes(f) for v in a) || error("keys don't match")
    map(i -> [v[i] for v in a], keys(f))
end
```

Even though I'm sure it's been tested thoroughly in SplitApplyCombine.jl, I'd rather keep a simple implementation here.

Reply:
Agreed